Matrix Exponentials for Sparse Matrices.

prashanth018 opened the pull request:
-- Making Mat Expo compatible for sparse matrix.
-- Added matrix norm estimating method majorly useful for estimating norm of sparse matrices.
-- Added helper methods (max, sum of matrices along any axes & zeros and ones matrices) that could be used generically by LA module.

TODO: Need to handle few edge cases.

Signed-off-by: Prashanth Duvvuri <prashanthsedan@gmail.com>

npadmana commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -2490,6 +2490,151 @@ proc cosm(A: []) throws {
   }
 }
 
+proc sum(A: [?D], param axis=0) where !isCSDom(D) {
```
minor nit -- I might rename this `absSum` (or something like that) as opposed to `sum`. Especially if this is user-facing.

npadmana replied:
A few other design questions -
1. Should this throw if `axis!=0,1`?
2. Should this be more general and handle summing over any subset of indices?

prashanth018 replied:
Resolved 1, working on 2.

npadmana commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -2490,6 +2490,151 @@ proc cosm(A: []) throws {
+    return forall j in D.dim(1-axis) do (+ reduce abs(A[D.dim(axis), j..j]));
+  else if axis == 1 then
+    // axis=1 sums along row
+    return forall i in D.dim(1-axis) do (+ reduce abs(A[i..i, D.dim(axis)]));
```
I'm guessing that the `forall` here should be `D.dim(axis)`.

npadmana commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -2490,6 +2490,151 @@ proc cosm(A: []) throws {
+  if axis == 0 {
+    // axis=0 sums along column
+    var b: [0..<t] A.eltType;
+    forall (i,j) in sps {
```
This looks like a race condition on `b[j]` (and same below). You will need to add a `with (+ reduce b)` clause to the `forall` loop.

npadmana commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -2490,6 +2490,151 @@ proc cosm(A: []) throws {
+
+proc ones(Dom: domain(2), type eltType=real) {
+  const (m,n) = Dom.shape;
+  var A: [Dom] eltType;
```
Can we simplify this to 
`var A: [Dom] eltType = 1 : eltType`?

prashanth018 replied:
Oh wow, Nice! So `1 : eltType` is like typecasting 1 to eltType ?

npadmana reviewed (commented):
I know you probably have a more complete PR, but here are some thoughts as I looked over this.

e-kayrakli commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -2490,6 +2490,151 @@ proc cosm(A: []) throws {
+
+proc sum(A: [?ADom], param axis=0) where isCSDom(ADom) {
+  var (n,t) = ADom.shape;
+  var sps: sparse subdomain(ADom.parentDom);
```
Is there any particular reason we need `sps`? I think it is identical to `ADom`

npadmana commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -2490,6 +2490,270 @@ proc cosm(A: []) throws {
+
+proc sign_round_up(A: [?D]){
+  forall (i,j) in D {
+    if b[i,j] == 0.0 then {
```
is `b[i,j]` meant to be `A[i,j]`?

If so, there is no reason to loop over indices -- you can just loop over the array elements
```
forall a in A {
  <change a>
}
```
@lydia-duncan or @e-kayrakli might have even cleaner approaches.

npadmana commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -2490,6 +2490,270 @@ proc cosm(A: []) throws {
+    // axis=0 sums along column
+    var b: [0..<t] A.eltType;
+    forall (i,j) in sps {
+      b[j] += A[i,j];
```
Here and below, is this meant to be `abs(A[i,j])`?

prashanth018 replied:
Yes! Thanks for that. I had it right in absSum method for rectangular matrix.

prashanth018 commented:
Hi All, Sorry that this comment has come in late. Below are a few of the NumPy-oriented methods that I have added to LA.chpl..

```
sparseDenseMatmul: Mutliplies Sparse and Dense Matrix
sign_round_up: Rounds up the elements of the Matrix to 1.0 if element >= 0, -1.0 otherwise
argsort: Given an array A, returns indexes corresponding to the sorted ordering of A.
elementary_vector: Returns an index's unit vector of size n.
zeros: Given the Domain and element-type, method returns a Matrix of zeros.
ones: Given the Domain and element-type, method returns a Matrix of ones.
max: Gets max along specified axes. (Method for Sparse matrix)
absSum: Gets sums along specified axes. (Added Methods for both Sparse and Dense matrix)
```
 Also added few other utility methods to support the implementation of `onenormest` of Sparse matrix.
 
Now that I'm doing this, please do let me know if there are any other methods (around these methods) that I can include to improve the LA module.

npadmana commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -2490,6 +2490,372 @@ proc cosm(A: []) throws {
+    throw new LinearAlgebraError("absSum(): axis=0 or axis=1 expected");
+  }
+
+  if axis == 0 then
```
<note -- I just deleted my last comment, the code was fine>

Is there a reason to write this as `D.dim(1-axis)` instead of just writing it as `D.dim(0)` and `D.dim(1)`? That is easier to read (at least for me, although others might have a different opinion).

Also, I think you can do `A[.., j]` instead of `A[D.dim(axis),j..j]`.

prashanth018 replied:
Sure Nikhil, I'll check that. I think there was a reason why I did that, I should have added a comment for that, I'm a bit out of context now since I have added it a while ago, I'll reiterate this method and let you know :)

npadmana commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -2490,6 +2490,372 @@ proc cosm(A: []) throws {
+  var (n,t) = ADom.shape;
+  if axis == 0 {
+    // axis=0 sums along column
+    var b: [0..<t] A.eltType;
```
Again, I think this is a race on b. I think you need a `with (max reduce b)` here as well.

prashanth018 replied:
Ah! I had this on the checklist before pushing. I missed it! I'll add this.

npadmana commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -2490,6 +2490,372 @@ proc cosm(A: []) throws {
+// Given the Domain and element-type, method
+// returns a Matrix of ones.
+proc ones(Dom: domain(2), type eltType=real) {
+  const (m,n) = Dom.shape;
```
I don't think you need this `const (m,n) ` line.

npadmana replied:
And the same with `zeros` below.

prashanth018 commented:
Created a new PR: [18149](https://github.com/chapel-lang/chapel/pull/18149)

npadmana commented:
Could you remind me of the status of this PR? I'm not sure if its been absorbed by others or if it is waiting for something (possibly me?)? Thanks!

e-kayrakli commented:
> Could you remind me of the status of this PR? I'm not sure if its been absorbed by others or if it is waiting for something (possibly me?)? Thanks!


I think the new PR has the commits here. I am closing this based on the silence. If I misinterpreted something, feel free to reopen it.
